File: src/Graph.cpp
#include "Graph.h"



Grafo::Grafo(int numConflictos, const std::vector<std::tuple<int, int>>& conflictos) {
    numNodos = 0;
    
    for (const auto& [u, v] : conflictos) {
        numNodos = std::max({numNodos, u, v});
    }
    numNodos++; 

    adjList.resize(numNodos);

    for (const auto& [u, v] : conflictos) {
        agregarArista(u, v);
    }
}

void Grafo:: agregarArista(int u, int v) {
    adjList[u].push_back(v);
    adjList[v].push_back(u);
}

bool Grafo:: sonAdyacentes(int u, int v) const {
    for (int vecino : adjList[u]) {
        if (vecino == v) return true;
    }
    return false;
}

void Grafo::addConflict(int item1, int item2) {
    if (!sonAdyacentes(item1, item2)) {
        agregarArista(item1, item2);
    }
}


void Grafo::removeConflict(int item1, int item2) {
    auto& adjacents1 = adjList[item1];
    auto& adjacents2 = adjList[item2];

    adjacents1.erase(std::remove(adjacents1.begin(), adjacents1.end(), item2), adjacents1.end());
    adjacents2.erase(std::remove(adjacents2.begin(), adjacents2.end(), item1), adjacents2.end());
}

File: python/Graph.py
class Graph:
    def __init__(self, num_conflicts=0, conflicts=None):
        self.num_nodes = 0
        self.adj_list = []
        
        if conflicts:
            # Find the maximum node index to determine the number of nodes
            for u, v in conflicts:
                self.num_nodes = max(self.num_nodes, u, v)
            self.num_nodes += 1  # Adjust for 0-indexing
            
            # Initialize adjacency list
            self.adj_list = [[] for _ in range(self.num_nodes)]
            
            # Add all conflict edges
            for u, v in conflicts:
                self.add_edge(u, v)
    
    def add_edge(self, u, v):
        """Add an edge between nodes u and v."""
        self.adj_list[u].append(v)
        self.adj_list[v].append(u)
    
    def are_adjacent(self, u, v):
        """Check if nodes u and v are adjacent."""
        return v in self.adj_list[u]
    
    def add_conflict(self, item1, item2):
        """Add a conflict between two items if it doesn't exist."""
        if not self.are_adjacent(item1, item2):
            self.add_edge(item1, item2)
    
    def remove_conflict(self, item1, item2):
        """Remove a conflict between two items."""
        if item2 in self.adj_list[item1]:
            self.adj_list[item1].remove(item2)
        if item1 in self.adj_list[item2]:
            self.adj_list[item2].remove(item1)